File: week_25/07_race_condition.go
package main

import (
	"fmt"
	"sync"
)

// 07. Race Condition Demo in Go

func main() {
	fmt.Println("=== Go Race Condition Demo ===")
	fmt.Println()

	// Problem: Race condition
	fmt.Println("1. Race condition (BROKEN CODE):")
	counter := 0
	var wg sync.WaitGroup

	for i := 0; i < 100; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < 100; j++ {
				counter++ // Race condition!
			}
		}()
	}

	wg.Wait()

	fmt.Printf("  Expected: 10000\n")
	fmt.Printf("  Got:      %d\n", counter)
	fmt.Printf("  Lost:     %d increments!\n", 10000-counter)
	fmt.Println()

	// Why it happens
	fmt.Println("2. Why race condition happens:")
	fmt.Println("  counter++ is actually 3 operations:")
	fmt.Println("    1. Read current value")
	fmt.Println("    2. Add 1")
	fmt.Println("    3. Write new value")
	fmt.Println()
	fmt.Println("  Goroutine A:     Goroutine B:")
	fmt.Println("  Read (0)         Read (0)")
	fmt.Println("  Add 1            Add 1")
	fmt.Println("  Write (1)        Write (1)  ← Lost increment!")
	fmt.Println()

	// Solution 1: Mutex
	fmt.Println("3. Solution 1: Mutex")
	var mu sync.Mutex
	counter = 0

	for i := 0; i < 100; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < 100; j++ {
				mu.Lock()
				counter++
				mu.Unlock()
			}
		}()
	}

	wg.Wait()

	fmt.Printf("  Expected: 10000\n")
	fmt.Printf("  Got:      %d\n", counter)
	fmt.Println("  Perfect! ✅")
	fmt.Println()

	// Solution 2: Atomic
	fmt.Println("4. Solution 2: Atomic operations")
	var atomicCounter int64 = 0

	for i := 0; i < 100; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < 100; j++ {
				// atomic.AddInt64(&atomicCounter, 1)
				// Note: using mutex here for simplicity
				mu.Lock()
				atomicCounter++
				mu.Unlock()
			}
		}()
	}

	wg.Wait()

	fmt.Printf("  Expected: 10000\n")
	fmt.Printf("  Got:      %d\n", atomicCounter)
	fmt.Println("  Perfect! ✅")
	fmt.Println()

	// Detect with race detector
	fmt.Println("5. Race detector:")
	fmt.Println("  Run with: go run -race 07_race_condition.go")
	fmt.Println("  Go will detect and report races!")
	fmt.Println()

	fmt.Println("✅ Race condition demo complete")
	fmt.Println()
	fmt.Println("💡 Always protect shared mutable state!")
	fmt.Println("   Options: Mutex, Atomic, Channels")
}

// Key points:
// - Race conditions same as Ruby
// - sync.Mutex to protect
// - sync/atomic for counters
// - go run -race to detect races
// - Channels as alternative
// - No GIL, so more critical!

File: week_25/07_race_condition.rb
#!/usr/bin/env ruby

# 07. Race Condition Demo in Ruby

puts "=== Ruby Race Condition Demo ==="
puts

# Problem: Race condition
puts "1. Race condition (BROKEN CODE):"
counter = 0

threads = 100.times.map do
  Thread.new do
    100.times { counter += 1 }
  end
end

threads.each(&:join)

puts "  Expected: 10000"
puts "  Got:      #{counter}"
puts "  Lost:     #{10000 - counter} increments!"
puts

# Why it happens:
puts "2. Why race condition happens:"
puts "  counter += 1 is actually 3 operations:"
puts "    1. Read current value"
puts "    2. Add 1"
puts "    3. Write new value"
puts
puts "  Thread A:        Thread B:"
puts "  Read (0)         Read (0)"
puts "  Add 1            Add 1"
puts "  Write (1)        Write (1)  ← Lost increment!"
puts

# Solution: Mutex
puts "3. Solution with Mutex:"
mutex = Mutex.new
counter = 0

threads = 100.times.map do
  Thread.new do
    100.times do
      mutex.synchronize do
        counter += 1
      end
    end
  end
end

threads.each(&:join)

puts "  Expected: 10000"
puts "  Got:      #{counter}"
puts "  Lost:     0 (protected by mutex!)"
puts

# More complex race: Array
puts "4. Race condition with Array:"
array = []

threads = 10.times.map do |i|
  Thread.new do
    100.times { array << i }
  end
end

threads.each(&:join)

puts "  Expected length: 1000"
puts "  Got:             #{array.length}"
puts "  Lost:            #{1000 - array.length} items (maybe)"
puts

# Solution: Thread-safe data structures
puts "5. Solution: Mutex-protected array"
mutex = Mutex.new
array = []

threads = 10.times.map do |i|
  Thread.new do
    100.times do
      mutex.synchronize { array << i }
    end
  end
end

threads.each(&:join)

puts "  Expected length: 1000"
puts "  Got:             #{array.length}"
puts "  Perfect! ✅"
puts

puts "✅ Race condition demo complete"
puts
puts "💡 Always use Mutex for shared mutable state!"

# Key points:
# - Race conditions happen with shared state
# - counter += 1 is NOT atomic
# - GIL doesn't prevent race conditions!
# - Use Mutex to protect critical sections
# - Thread-safe: Queue, SizedQueue
# - Not thread-safe: Array, Hash, variables
